### PR TITLE
Use vispy for plotting backend

### PR DESCRIPTION
**Pull request by dstansby**
This changes the plotting backend from `pyqtgraph` to `vispy`. Currently a work in progress, but I'm opening a draft PR to mark that the work is being done! Once this is ready I'll take it out of draft status and request a review.

Fixes https://github.com/lowe-lab-ucl/arboretum/issues/34, fixes https://github.com/lowe-lab-ucl/arboretum/issues/29

### CODE REVIEW

**Comment by quantumjot** [comment]
This is great! I think we should consider pushing an updated version to the hub/PyPI once this is merged. It would be good to let some other users try it and give some feedback too

File: examples/show_sample_data.py [hunk 1]
```diff
@@ -14,8 +14,8 @@
 track, segmentation = load_sample_data()
 
 viewer = napari.Viewer()
-viewer.add_layer(track)
 viewer.add_layer(segmentation)
+viewer.add_layer(track)
```
Comment (paddyroddy) [inline examples/show_sample_data.py:18]: Out of interest, is there a reason for the swap?

File: napari_arboretum/visualisation/vispy_plotter.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    def get_qwidget(self) -> QWidget:
+        return self.canvas.native
+
+    def clear(self) -> None:
+        self.tree.clear()
+
+    def autoscale_view(self) -> None:
+        """Scale the canvas so all branches are in view."""
+        xs = np.concatenate([track.pos[:, 0] for id, track in self.tree.tracks.items()])
+        ys = np.concatenate([track.pos[:, 1] for id, track in self.tree.tracks.items()])
```
Comment (paddyroddy) [inline napari_arboretum/visualisation/vispy_plotter.py:44]: I suspect these would work as generators? As in, you may be able to remove the `[]`

File: napari_arboretum/visualisation/vispy_plotter.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        """
+        Add a single label to the tree.
+        """
+        self.tree.add_annotation(a.x, a.y, a.label, a.color)
+
+    def set_title(self, title: str) -> None:
+        """
+        Set the title of the plot.
+        """
+        pass
```
Comment (paddyroddy) [inline napari_arboretum/visualisation/vispy_plotter.py:73]: Is this intentional?

File: napari_arboretum/visualisation/vispy_plotter.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+
+class TreeVisual(scene.visuals.Compound):
+    """
+    Tree visual that stores branches as sub-visuals.
+    """
+
+    def __init__(self, parent):
+        super().__init__([])
```
Comment (paddyroddy) [inline napari_arboretum/visualisation/vispy_plotter.py:82]: Why the `[]`?

File: napari_arboretum/tree.py [hunk 5]
```diff
@@ -68,25 +64,15 @@ def _build_tree(nodes) -> Tuple[List[Edge], List[Annotation]]:
         node = queue.pop(0)
         y = y_pos.pop(0)
 
-        # TODO(arl): sync this with layer coloring
```
Comment (paddyroddy) [inline napari_arboretum/tree.py:71]: I assume this is completed?

**Review by paddyroddy — APPROVED** [review]
Looks good

File: examples/show_sample_data.py (hunk 1 above)
Comment (dstansby) [reply examples/show_sample_data.py:18]: This way round the track layer is the one selected when the `napari` viewer opens, which means you can click on it straight away to pull up `arboretum`.

File: napari_arboretum/visualisation/vispy_plotter.py (hunk 4 above)
Comment (dstansby) [reply napari_arboretum/visualisation/vispy_plotter.py:82]: `scene.visuals.Compound` needs a list of sub-visuals to initialise with.

Comment (dstansby) [reply napari_arboretum/visualisation/vispy_plotter.py:82]: (even if it's an empty list)

File: napari_arboretum/visualisation/vispy_plotter.py (hunk 3 above)
Comment (paddyroddy) [reply napari_arboretum/visualisation/vispy_plotter.py]: Sounds sensible to me

**Pull request merged**